### ISSUE TITLE
running pypyjs inside Worker('mtpypy.js') crashes on extra message events

### ISSUE THREAD

**Issue by perkinslr** [rfk/pypyjs#67]
I pushed the vm handling code to a separate file (mtpypy.js), which I include in the index.html via 
  worker=new Worker('mtpypy.js');

Then the prompt function sends the user input into the worker thread, and stdout sends output from the worker thread to the display.  All that works, and lets you run long-running code without freezing the UI.  However, if messages get posted to the worker while the worker is not waiting for user input, it sometimes crashes via os.abort().  

This happens as well with WebSocket .onmessage events, and I suspect will happen with any asynchronous callbacks which can interrupt the vm.

**Comment by perkinslr** [comment]
The tracebacks I get are here
https://gist.github.com/perkinslr/e6f04a0ea69f37f4d17c

Happens in both firefox and chrome, the traceback is from chrome.

**Comment by perkinslr** [comment]
Well, I should have done some more poking before posting this.  
It is specifically when python callbacks fire from JS while the VM is spinning that causes the crash.  I can store the values inside the worker thread and pull them in from python as needed, pushing them to the python via callback is what breaks.

To reproduce, create a websocket factory and call it from in python, then specify the ws.onmessage=lambda x: print(x)
then connect to an echo server, send a bunch of messages in a loop.  The next time you try to eval code, it should crash.

**Comment by rfk** [comment]
Interesting.  If possible (just to make it easier for me to dive in) can you please package a testcase up as a tarball and I'll try to reproduce and dig into this?

There's some weirdness with callbacks and garbage-collection currently, which I imagine will be particularly bad if you're using `lambda`; see e.g. #22.  You need to somehow keep the function alive on the python side after sending it off to javascript, since we can't do cross-language cycle detection.   I wonder if that's what is causing the issue here.

**Comment by perkinslr** [comment]
I believe that is the problem.  I've encountered similar problems with cffi in pypy in the past, so I probably should have thought of it on my own.  Even in my real code, I was using a closure for the callback, and not holding any references to it.  I changed it to hold both a reference to the python function and the js wrapper and the problem appears to be solved.  

If you want, I can still provide a tarball with a minimal example.  Otherwise I'm getting ready to push my full version to my fork of the project; the _socket module is where I first encountered this bug.
